Support ERC20 Token Currency

nkoljanin opened the pull request:
Enables purchasing of the booking token with ERC20 Token Support.
Added checks for allowance, and approving if not enough allowance.
Handles the ERC20 Token support in the bot and in the examples.
Added a payment config to the mint examples to choose between 3 ways of payment (token, native token, iso currency)

Tested through Postman and mintnbuy.go script

evlekht commented on pkg/booking/booking.go:
```diff
@@ -86,7 +91,16 @@ func (bs *Service) MintBookingToken(
 	}
-
+	if paymentToken != zeroAddress {
+		erc20Contract, err := erc20.NewErc20(paymentToken, bs.client)
```
I'd suggest to add lru cache for bindings, there probably will be just a few ones that will be used often. Create new struct that will contain that cache, init it in app and pass this struct to both booking and responseHandler

evlekht commented on pkg/booking/booking.go:
```diff
@@ -57,11 +61,12 @@ func NewService(
+		cmAccount:        cmAccount,
+		transactOpts:     transactOpts,
+		chainID:          chainID,
+		cmAccountAddress: &cmAccountAddr,
```
I'd suggest to store addr as value, not as pointer - will be more convenient to use, since u'r passing it as value in func arg anyway

nkoljanin committed:
rebased

add tokenCache to mintbuy example
linter fixes
wip fixes for token cache
token cache wip
Adding token cache
storing addr as value
go mod
Merge branch 'dev' into nkoljanin/erc20_support
Package import
fixes
renaming
ERC20 Token Payment Support
Changed host for camino also to messenger.chain4travel.com in example config (#55)

Co-authored-by: Noctunus <alexander.koch@chain4travel.com>
Fixing ordering of processing in auto-generation of services to make sure that on every system we get the same result. (#54)

Co-authored-by: Noctunus <alexander.koch@chain4travel.com>
Add mint v1 support (#53)
wip - testing
ERC20 token support

evlekht committed:
fixes

evlekht committed:
fixes

evlekht committed:
fixes
